Sources/JWT/JWKSCache.swift

```diff
@@ -3,84 +3,137 @@ import Vapor
 /// A thread-safe and atomic class for retrieving JSON Web Key Sets which honors the
 /// HTTP `Cache-Control`, `Expires` and `Etag` headers.
 public final class JWKSCache {
+    public enum Error: Swift.Error {
+        case missingCache
+        case unexpctedResponseStatus(HTTPStatus, uri: URI)
+    }
     private let uri: URI
+    private let client: Client
+    private let sync: Lock
 
-    // Uses a private event loop so that read of the cache date and the possible
-    // subsequent network download happens atomically.  Don't want multiple requests
-    // coming in at once and kicking off multiple network downloads.
-    private let eventLoop: EventLoop
+    struct CachedJWKS {
+        var cacheUntil: Date
+        var jwks: JWKS
+        var eTag: String?
+    }
 
-    internal var cacheUntil: Date?
-    internal var jwks: JWKS?
-    internal var etag: String?
+    private var cachedJWKS: CachedJWKS?
+    private var currentRequest: EventLoopFuture<JWKS>?
 
-    /// The initializer.
+    /// Creates a new `JWKSCache`.
     /// - Parameters:
     ///   - keyURL: The URL to the JWKS data.
     ///   - application: The Vapor `Application`.
-    public init(keyURL: String, on application: Application) {
+    public init(keyURL: String, client: Client) {
```

Comment:
Should we provide another initializer that takes a Request?  That's what most people are going to have when they do this.  Just prevents them from doing `req.client`, which not everyone is going to realize exists.

Reply:
I wanted to allow passing `Client` here since you might need to create a one-off client that is capable of fulfilling the request. In the test for example, we have to disable unclean shutdown. You might want to create a separate client with that configuration just for this purpose.